File: javascript/rotational-cipher/rotational-cipher.js
function rotate(char, shift) {
    if (!char.match(/[a-z]/i)) {
        return char;
    }

    const code = char.charCodeAt() + shift;
    const base = (char.match(/[a-z]/) ? 'a' : 'A').charCodeAt();
    const limit = (char.match(/[a-z]/) ? 'z' : 'Z').charCodeAt();

    if (code <= limit) {
        return String.fromCharCode(code);
    }

    return String.fromCharCode(code + base - limit - 1);
}

export class RotationalCipher {
    static rotate(string, shift) {
        return [...string].map(char => rotate(char, shift)).join('');
    }
}

File: java/rotational-cipher/src/main/java/RotationalCipher.java
class RotationalCipher {

    private int shift;

    RotationalCipher(int shift) {
        this.shift = Math.abs(shift % 26);
    }

    public String rotate(String data) {
        return data.codePoints()
            .mapToObj(this::rotate)
            .collect(StringBuilder::new,
                     StringBuilder::appendCodePoint,
                     StringBuilder::append)
            .toString();
    }

    private int rotate(int code) {
        if (isLetter(code)) {
            char base = Character.isUpperCase(code) ? 'A' : 'a';
            return ((code - base + shift) % 26) + base;
        }

        return code;
    }

    private boolean isLetter(int code) {
        return (code >= 'a' && code <= 'z')
            || (code >= 'A' && code <= 'Z');
    }
}
